File: fila/Fila Estática/Java/Fila Com Pilhas/FilaEstaticaComPilhas.java
public class FilaEstaticaComPilhas implements Enfileiravel {

	private Empilhavel p1;
	private Empilhavel p2;
	
	public FilaEstaticaComPilhas() {
		this(10);
	}
	
	public FilaEstaticaComPilhas(int tamanho) {
		p1 = new PilhaEstatica(tamanho);
		p2 = new PilhaEstatica(tamanho);
	}
	
	@Override
	public void enfileirar(Object dado) {
		if (!estaCheia()) {
			//já colocar os elementos como numa fila.. FIFO
			while(!p1.estaVazia()) {
				p2.empilhar(p1.desempilhar());
			}
			
			p1.empilhar(dado);

			while(!p2.estaVazia()) {
				p1.empilhar(p2.desempilhar());
			}			
		} else {
			System.err.println("Fila Cheia!");
		}
	}
	
	@Override	
	public Object desenfileirar() {
		Object dadoInicio = null;	
		if (!estaVazia()) {
			dadoInicio = p1.desempilhar();
		} else {
			System.err.println("Fila Vazia!");
		}
		return dadoInicio;
	}

	@Override
	public void atualizarInicio(Object dado) {
	}

	@Override
	public void atualizarFim(Object dado) {
	}
	
	@Override	
	public Object frente() {
		Object dadoInicio = null;
		if (!estaVazia()) {
			dadoInicio = p1.espiar();
		} else {
			System.err.println("Fila Vazia!");		
		}
		return dadoInicio;
	}
	
	@Override	
	public boolean estaCheia() {
		return p1.estaCheia();
	}
	
	@Override	
	public boolean estaVazia() {
		return p1.estaVazia();
	}
	
	@Override	
	public String imprimir() {
		return p1.imprimir();	
	}
}

File: fila/estatica/Kotlin/Fila com Pilhas/FilaEstaticaComPilhas.kt
class FilaEstaticaComPilhas(val tamanho: Int = 10) : Enfileiravel {
 
	private var pilha1: Empilhavel = PilhaEstatica(tamanho) 
	private var pilha2: Empilhavel = PilhaEstatica(tamanho)

	override fun enfileirar(dado: Any?) {
        //coloca dado na pilha2
        pilha2.empilhar(dado)
        //volta dados da pilha1 para pilha2
        while(!pilha1.estaVazia()) {
            pilha2.empilhar(pilha1.desempilhar())
        }
        //volta todos os dados pra pilha1
        while(!pilha2.estaVazia()) {
            pilha1.empilhar(pilha2.desempilhar())
        }        
    }
	
    override fun atualizar(dado: Any?) {
        pilha1.atualizar(dado)
    }

	override fun desenfileirar(): Any? {
        return pilha1.desempilhar()
    }
	
	override fun frente(): Any? {
        return pilha1.espiar()
    }
	
	override fun estaCheia(): Boolean {
        return pilha1.estaCheia()
    }
	
	override fun estaVazia(): Boolean {
        return pilha1.estaVazia()
	}
	
	override fun imprimir(): String {
        return pilha1.imprimir()
    }
}